src/beman/optional/tests/optional_range_support.t.cpp

```diff
@@ -0,0 +1,360 @@
+    ASSERT_FALSE(empty.has_value());
+
+    for (auto _ : empty) {
+        ASSERT_TRUE(false) << "Should not be reached: expected not to loop over empty optional";
+    }
+}
+
+TEST(RangeSupportTest, LoopOverNonEmptyRange) {
+    const auto                     expected_value = 0xCAFEBABE;
+    beman::optional::optional<int> empty{expected_value};
+    ASSERT_TRUE(empty.has_value());
+
+    bool entered_loop = false;
+    for (auto i : empty) {
+        EXPECT_EQ(i, expected_value);
+        entered_loop = true;
+    }
+    EXPECT_TRUE(entered_loop);
+}
+
+TEST(RangeSupportTest, LoopOptionalAccess) {
+    // Example from P3168R1: should access the value from an optional object.
+    const int  expected_value = 0xCAFEBABE;
+    const auto get_optional   = [&]() -> beman::optional::optional<int> { return expected_value; };
+    ASSERT_TRUE(get_optional().has_value());
+
+    for (auto&& opt : get_optional()) {
+        EXPECT_EQ(opt, expected_value); // usage of opt here is safe
+    }
+}
+
+TEST(RangeSupportTest, LoopOptionalAssignment) {
+    // Example from P3168R1: should mutate the value from an optional object.
+    const int  initial_expected_value = 0xCAFEBABE;
+    const int  expected_value         = 0xDEADBEEF;
+    const auto get_optional           = [&]() -> beman::optional::optional<int> { return initial_expected_value; };
+    ASSERT_TRUE(get_optional().has_value());
+    ASSERT_TRUE(get_optional().value() == initial_expected_value);
+
+    auto opt_int = get_optional();
+    for (auto&& opt : opt_int) {
+        EXPECT_EQ(opt, initial_expected_value);
+        opt = expected_value; // usage of opt here is safe
+    }
+    ASSERT_TRUE(opt_int.has_value());
+    EXPECT_EQ(opt_int.value(), expected_value);
+}
+
+TEST(RangeSupportTest, RangeChainExampleWithVector) {
+    // Example from P3168R1: start from a vector of values, apply multiple range operations involving optional values.
+    std::vector<int> v{2, 3, 4, 5, 6, 7, 8, 9, 1};
+    const auto       test = [](int i) -> beman::optional::optional<int> {
+        switch (i) {
+        case 1:
+        case 3:
+        case 7:
+        case 9:
+            return i;
+        default:
+            return {};
+        }
+    };
+
+    auto&& r =
+        v                                     // starting vector
+        | std::ranges::views::transform(test) // generates {nullopt, 3, nullopt, 7, nullopt, nullopt, nullopt, 9, 1}
+        | std::ranges::views::filter([](auto x) {
+              return bool(x);
+          }) // generates {optional<int>{3}, optional<int>{7}, optional<int>{9}, optional<int>{1}}
+        | std::ranges::views::transform([](auto x) { return *x; }) // generates {3, 5, 7, 9}
+        | std::ranges::views::transform([](int i) {
+              // std::cout << i << "\n"; // do not actually do it in tests
+              return i;
+          }) // print + identity transform
+        ;
+
+    ASSERT_TRUE(std::ranges::equal(r, std::vector<int>{3, 7, 9, 1}));
```

Comment:
This doesn't doesn't actually rely on `optional` being a range.

Reply:
Right, thanks!

Reply:
Right, here we would just skip a transform operation. I removed the test.